### PR TITLE
[Experimental] Introduce visibility to BUILD file

### PR DESCRIPTION
**Pull request by yang-g**
Worth a try?

gRPC top level BUILD file has the following issues because of lack of visibility:

- It is not clear which targets are for users to depend on because all the targets are public.
- No way to hide internal headers from users, thus any change can potentially break users depending on internal libraryies.

### CODE REVIEW

**Comment by jtattermusch** [comment]
Before sending for review, please include some description of the benfits of the change.

**Comment by yang-g** [comment]
> Before sending for review, please include some description of the benfits of the change.

done

File: BUILD [hunk 1]
```diff
@@ -301,6 +301,7 @@ grpc_cc_library(
     language = "c++",
     public_hdrs = GPR_PUBLIC_HDRS,
     standalone = True,
+    visibility = ["//visibility:public"],
```
Comment (jtattermusch) [inline BUILD:304]: how were the list of "public" libraries established? Did you just hand-picked those? How can we make sure the list is complete and there's nothing missing?

AFAIK not all the public libraries are actually in the /BUILD file, there are also some public items in some of the subdirectories.

Comment (yang-g) [reply BUILD:304]: I marked all the standalone targets public by hand. Note this PR is just dealing with the top level BUILD file, and other targets in subdirectories are out of scope.

Comment (jtattermusch) [reply BUILD:304]: I think we need to make sure that there's a clear mapping between `visibility = ["//visibility:public"]` and the `build` attribute in the script that generates build.yaml metadata from the bazel build:
https://github.com/grpc/grpc/blob/a43082112f5a2e4ef475c8b9c364ca08349e201b/tools/buildgen/extract_metadata_from_bazel_xml.py#L575
(there's clearly some relationship between the "public" targets and the ones marked as e.g. `build: all` and while doing this cleanup we should figure out what the source of truth is).

**Pull request closed without merging**